types/meteor/hot-module-replacement.d.ts

```diff
@@ -0,0 +1,20 @@
+interface HotModuleReplacementData {
+    // user-defined attributes
+}
+
+declare namespace NodeJS {
+    interface Module {
+        readonly hot?: {
+            accept(): void;
+            decline(): void;
+            dispose(callback: (data: HotModuleReplacementData) => void): void;
+            data: HotModuleReplacementData | null;
+            onRequire<T>(callbacks: {
+                before?(requiredModule: Module, parentId: string): T;
+                after?(requiredModule: Module, data: T): void;
+            }): void;
```

Comment:
I have a problem with the `HotModuleReplacementData` type. While adding the generic parameter to `onRequire` helps(*), introducing an interface for `data` is problematic. Interfaces in TypeScript are merged _globally_, not _per-module_. So the example from the tests you've added would make all of the modules to "see" this property (`color`) of only one of them. In other words: let's use `unknown` instead.

(*) It's not necessarily true that the type of data returned from `before` will match the one available in `after`, as it may change while changing the file. For example, if you'd remove a property, the "old" `after` function would receive the "new" data of a different shape. However, it's not something we should care about here, as types should reflect the static code.

Reply:
Good points. I wish there were a way to add `data` attributes locally to the module, but I don't know one.

I've gone ahead and removed `HotModuleReplacementData`. Instead of `unknown`, I used the slightly more precise `object`, because `dispose` is guaranteed to be given an `object` not `null`.